### PR TITLE
Added Fn2, Fn3, apply2, apply3

### PR DESCRIPTION
**Pull request by Icelandjack**
Discussion #110.

### CODE REVIEW

File: Test/QuickCheck/Function.hs [hunk 1]
```diff
@@ -452,20 +456,65 @@ shrinkFun shr (Map g h p) =
...
+-- @ 
+--     prop_zipWith :: Fun (Int, Bool) Char -> [Int] -> [Bool] -> Bool
+--     prop_zipWith (Fn2 f) xs ys = zipWith f xs ys == [ f x y | (x, y) <- zip xs ys]
+-- @
+--
+-- >>> quickCheck prop_zipWith
+-- +++ OK, passed 100 tests.
+-- 
+pattern Fn2 :: (a -> b -> c) -> Fun (a, b) c
+pattern Fn2 f <- Fun _ (curry -> f)
```
Comment (phadej) [inline Test/QuickCheck/Function.hs]: needs `ViewPatterns`

File: Test/QuickCheck/Function.hs [hunk 2]
```diff
@@ -452,20 +456,65 @@ shrinkFun shr (Map g h p) =
...
+-- 
+-- @ 
+--     prop_zipWith :: Fun (Int, Bool) Char -> [Int] -> [Bool] -> Bool
+--     prop_zipWith (Fn2 f) xs ys = zipWith f xs ys == [ f x y | (x, y) <- zip xs ys]
+-- @
+--
+-- >>> quickCheck prop_zipWith
+-- +++ OK, passed 100 tests.
+-- 
+pattern Fn2 :: (a -> b -> c) -> Fun (a, b) c
```
Comment (phadej) [inline Test/QuickCheck/Function.hs]: Type signatures for patterns works only with `GHC 7.10+` (it's omitted for `Fn` for that reason)

Comment (phadej) [inline Test/QuickCheck/Function.hs]: *FWIW* I'd add `type Fun2 a b = Fun (a, b)` and `type Fun3 a b c = Fun (a,b,c)` aliases (eta-reduced `c`)

Comment (Icelandjack) [inline Test/QuickCheck/Function.hs]: I carefully commented it out for that exact reason, I'm guess I'm a serial type signature added :)

I added your suggestions

**Pull request merged**